### PR TITLE
PP 59 My lessons - student dashboard

### PR DESCRIPTION
**Pull request by vrudnitskaya**
### Description
Implement displaying the next two lessons on the Student's home page. If there are less than 7 days left until the lesson, a timer is displayed.

The My Lessons page displays all the courses a student is attending. Each class has lessons attached to it. They are divided into upcoming and previous.

### Screenshots
home-student-nexttwo-lessons2
lessons-table
prev-lessons

### CODE REVIEW

File: src/components/pages/dashboard-pages/HomeStudent.jsx [hunk 1]
```diff
@@ -73,29 +80,78 @@ const HomeStudent = ({ profile, onNavigate, profileError }) => {
...
+                        src={lesson.teacherPhoto}
+                        alt="teacher photo"
+                        className="rounded-full w-20 h-20 object-cover"
+                      />
+                      <div className="flex flex-col justify-around w-2/3">
+                        <p className="font-medium">
+                          {lesson.teacherFirstName}{' '}
+                          {lesson.teacherLastName.slice(0, 1)}.
+                        </p>
+                        <p>{lesson.teacherCategory.join('& ')} teacher</p>
```
Comment (mariyador) [inline src/components/pages/dashboard-pages/HomeStudent.jsx:111]: Сould you please add a space before  & , to make teacher categories look better.

File: src/components/pages/dashboard-pages/Lessons.jsx [hunk 2]
```diff
@@ -1,11 +1,220 @@
...
+          <Link className="p-4 underline" to="/search">
+            Search classes or teachers
+          </Link>
+        </div>
+      )}
+      <div className="flex sm:flex-row flex-col gap-4 sm:gap-1 w-full justify-evenly md:p-4 items-start mb-10">
+        {studentClasses.length > 0 && (
+          <div className="bg-pureWhite w-10/12 sm:w-1/4 flex flex-col items-center self-center sm:self-start rounded">
+            <h1 className="text-black font-semibold text-xl font-spartan text-center py-4">
+              My lessons
```
Comment (mariyador) [inline src/components/pages/dashboard-pages/Lessons.jsx:138]: I think it is much better to name this section My classes

**Review by mariyador — APPROVED** [review]
Valya! 
You did such a great job. A add comments but everything looks good on my side!

Screenshot 2024-09-01 at 2 10 37 AM

Screenshot 2024-09-01 at 2 11 04 AM

File: src/components/pages/NotFound.jsx [hunk 3]
```diff
@@ -3,13 +3,13 @@ import img from '../../assets/imgPage404.jpg';
 
 const NotFound = () => {
   return (
-    <div className="bg-pureWhite    w-full h-screen flex flex-col gap-7 items-center mt-24">
+    <div className="bg-pureWhite w-full flex flex-col gap-7 items-center mt-4">
       <h1 className="text-red font-bold text-2xl sm:text-4xl font-spartan uppercase">
         Page not found
       </h1>
-      <img src={img} alt="Error 404" className="w-[350px] sm:w-[550px]" />
+      <img src={img} alt="Error 404" className="w-4/5 md:w-2/5" />
```
Comment (AnnaPestova1) [inline src/components/pages/NotFound.jsx:10]: it is good to have not hardcoded width, great!

File: src/components/pages/dashboard-pages/HomeStudent.jsx [hunk 4]
```diff
@@ -73,29 +80,78 @@ const HomeStudent = ({ profile, onNavigate, profileError }) => {
...
             Your upcoming lessons
           </h2>
-          <div>
-            <p className="px-4">
-              {myLessons.length > 0 ? myClasses : 'No lessons booked yet'}
-            </p>
-          </div>
+          {nextTwoLessons.length > 0 ? (
+            <div className="flex gap-8 flex-col lg:flex-row">
+              {nextTwoLessons.map((lesson) => {
```
Comment (AnnaPestova1) [inline src/components/pages/dashboard-pages/HomeStudent.jsx:89]: lines 89-138 has a long return in map function. We could move this logic in a separate variable or even component.

File: src/util/LessonsService.js [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+          lessonSchedule: {
+            ...lesson.lessonSchedule,
+            date: new Date(lesson.lessonSchedule.date),
+          },
+        }))
+        .filter(lesson => lesson.lessonSchedule.date >= today)
+        .sort((a, b) => a.lessonSchedule.date - b.lessonSchedule.date)
+        .slice(0, 2);
+
+      const nextLessonsWithDurationTitle = nextLessons.map(lesson => {
```
Comment (AnnaPestova1) [inline src/util/LessonsService.js:106]: I love this function

**Review by AnnaPestova1 — APPROVED** [review]
Hi Valya! Great work! 
Approved.

Just want to mention, that you have 2 components with a long jsx return that includes multiple conditional rendering, sometimes it is better to create additional components for a small chunk of logic to avoid complicated conditional code. This mention does not affect any piece of your great functionality and you shouldn't rewrite the code. 

I also realized that I want to see the date of all my classes here, do you think it is good addition and your team has time to implement this changes? 
image

File: src/components/pages/dashboard-pages/Lessons.jsx [hunk 6]
```diff
@@ -1,11 +1,220 @@
...
+                <button className="w-full sm:w-1/2 lg:w-full bg-red hover:bg-pureWhite hover:text-red h-10 hover:border-2 hover:border-red text-white font-spartan font-semibold md:text-xl rounded-lg transition duration-300 easy-in">
+                  Send message
+                </button>
+              </div>
+            </div>
+            <div className="self-start w-full">
+              <p className="lg:text-lg px-6">{selectedClass[0].description}</p>
+            </div>
+
+            {/* <div className='self-start'>
```
Comment (AnnaPestova1) [inline src/components/pages/dashboard-pages/Lessons.jsx:171]: do we need this code?

**Comment by vrudnitskaya** [comment]
> Hi Valya! Great work! Approved.
> 
> Just want to mention, that you have 2 components with a long jsx return that includes multiple conditional rendering, sometimes it is better to create additional components for a small chunk of logic to avoid complicated conditional code. This mention does not affect any piece of your great functionality and you shouldn't rewrite the code.
> 
> I also realized that I want to see the date of all my classes here, do you think it is good addition and your team has time to implement this changes (of course not in this PR)? image

Thank you, Anna!
Did you mean the date in the "My Classes" section of the teacher dashboard?

File: src/components/pages/dashboard-pages/Lessons.jsx (hunk 6 above)
Comment (vrudnitskaya) [reply src/components/pages/dashboard-pages/Lessons.jsx:171]: We haven't decided yet if we'll have time to implement homework upload. This is a draft for this case. But it's probably best to delete it.

File: src/components/pages/dashboard-pages/HomeStudent.jsx (hunk 4 above)
Comment (vrudnitskaya) [reply src/components/pages/dashboard-pages/HomeStudent.jsx:89]: I'll think how to do this.

**Pull request merged**